plugins/inputs/kinesis/kinesis.go

```diff
@@ -0,0 +1,337 @@
+	client := kinesis.New(configProvider)
+
+	k.checkpoint = &noopCheckpoint{}
+	if k.DynamoDB != nil {
+		var err error
+		k.checkpoint, err = ddb.New(
+			k.DynamoDB.AppName,
+			k.DynamoDB.TableName,
+			ddb.WithDynamoClient(dynamodb.New((&internalaws.CredentialConfig{
+				Region:      k.Region,
+				AccessKey:   k.AccessKey,
+				SecretKey:   k.SecretKey,
+				RoleARN:     k.RoleARN,
+				Profile:     k.Profile,
+				Filename:    k.Filename,
+				Token:       k.Token,
+				EndpointURL: k.EndpointURL,
+			}).Credentials())),
+			ddb.WithMaxInterval(time.Second*10),
+		)
+		if err != nil {
+			return err
+		}
+	}
+
+	cons, err := consumer.New(
+		k.StreamName,
+		consumer.WithClient(client),
+		consumer.WithShardIteratorType(k.ShardIteratorType),
+		consumer.WithCheckpoint(k),
+	)
+	if err != nil {
+		return err
+	}
+
+	k.cons = cons
+
+	k.acc = ac.WithTracking(k.MaxUndeliveredMessages)
+	k.records = make(map[telegraf.TrackingID]string, k.MaxUndeliveredMessages)
+	k.checkpoints = make(map[string]checkpoint, k.MaxUndeliveredMessages)
+	k.sem = make(chan struct{}, k.MaxUndeliveredMessages)
+
+	k.ctx = context.Background()
+	k.ctx, k.cancel = context.WithCancel(k.ctx)
+
+	k.wg.Add(1)
+	go func() {
+		defer k.wg.Done()
+		err := k.cons.Scan(k.ctx, func(r *consumer.Record) consumer.ScanStatus {
+			select {
+			case <-k.ctx.Done():
+				return consumer.ScanStatus{Error: k.ctx.Err()}
+			case k.sem <- struct{}{}:
+				break
+			}
+			err := k.onMessage(k.acc, r)
+			if err != nil {
+				k.sem <- struct{}{}
+				return consumer.ScanStatus{Error: err}
+			}
+
+			return consumer.ScanStatus{}
+		})
+		if err != nil {
+			log.Printf("E! [inputs.kinesis_consumer] Scan encounterred an error - %s", err.Error())
+			k.cons = nil
+		}
+	}()
+
+	return nil
+}
+
+func (k *KinesisConsumer) Start(ac telegraf.Accumulator) error {
+	err := k.connect(ac)
+	if err != nil {
+		return err
+	}
+
+	k.wg.Add(1)
+	go func() {
+		defer k.wg.Done()
+		k.onDelivery()
+	}()
+
+	return nil
+}
+
+func (k *KinesisConsumer) onMessage(acc telegraf.TrackingAccumulator, r *consumer.Record) error {
+	metrics, err := k.parser.Parse(r.Data)
+	if err != nil {
+		return err
+	}
+
+	k.recordsTex.Lock()
+	id := acc.AddTrackingMetricGroup(metrics)
+	k.records[id] = *r.SequenceNumber
+	k.recordsTex.Unlock()
+
+	return nil
+}
+
+func (k *KinesisConsumer) onDelivery() {
+	for {
+		select {
+		case <-k.ctx.Done():
+			return
+		case info := <-k.acc.Delivered():
+			k.recordsTex.Lock()
+			sequenceNum, ok := k.records[info.ID()]
+			if !ok {
+				k.recordsTex.Unlock()
+				continue
+			}
+			<-k.sem
+			delete(k.records, info.ID())
+			k.recordsTex.Unlock()
+
+			if info.Delivered() {
+				k.checkpointTex.Lock()
+				chk, ok := k.checkpoints[sequenceNum]
+				if !ok {
+					k.checkpointTex.Unlock()
+					continue
+				}
+				delete(k.checkpoints, sequenceNum)
+				k.checkpointTex.Unlock()
+
+				// at least once
+				if sequenceNum > k.lastSeqNum {
```

Comment:
It does if a "batch" is being processed, which is almost always the case. Line 283 sets the last sequence number to the current one, so if the batch is processed from oldest (smaller sequence number) to newest, it will be true every time and the checkpoint would be set at the oldest metric; unless a record comes in out of order, in which, if it is older, the checkpoint would be set to that record.